<launch>
    <arg name="show_view" default="true" />
    <arg name="use_bag" default="false" />
    <arg name="bag" default="$(find detect)/outputs/bags/robot_20.bag" />

    <arg name="fps" default="1" />
    <arg name="device" default="cuda:0" />

    <!-- <arg name="finger_num" default="4" /> -->
    <arg name="finger_num" default="3" />
    <arg name="unit_angle" default="3" />
    <!-- <arg name="hand_radius_mm" default="152.5" /> -->
    <arg name="hand_radius_mm" default="100" />
    <arg name="finger_radius_mm" default="15" />
    <!-- <arg name="hand_mount_rotation" default="45" /> -->
    <!-- <arg name="hand_mount_rotation" default="60" /> -->
    <arg name="hand_mount_rotation" default="75" /> <!-- カメラの回転45° + 指の世界座標系y軸に対する角度30° -->
    <!-- <arg name="hand_mount_rotation" default="30" /> -->
    <!-- <arg name="hand_mount_rotation" default="0" /> -->

    <arg name="robot_name" default="myrobot" />
    <arg name="debug" default="false" />
    <arg name="visualize_only_best_cnd" default="false" />

    <!-- <arg name="used_camera" default="left_camera" />
    <arg name="image_topic" default="/$(arg robot_name)/$(arg used_camera)/color/image_raw" />
    <arg name="depth_topic" default="/$(arg robot_name)/$(arg used_camera)/aligned_depth_to_color/image_raw" /> -->
    <arg name="image_info_topic" default="/$(arg robot_name)/left_camera/aligned_depth_to_color/camera_info" />
    <!-- <arg name="depth_info_topic" default="/$(arg robot_name)/$(arg used_camera)/aligned_depth_to_color/camera_info" /> -->
    <!-- <arg name="points_topic" default="/$(arg robot_name)/$(arg used_camera)/depth/color/points" /> -->
    <arg name="seg_topic" default="/segmented_img" />
    <!-- <arg name="candidates_img_topic" default="/$(arg robot_name)/$(arg used_camera)/color/image_raw/candidates_img" /> -->


    <!-- for realsense-viewer bags -->
    <!-- <remap from="/device_0/sensor_1/Color_0/image/data" to="$(arg image_topic)" />
    <remap from="/device_0/sensor_1/Color_0/info/camera_info" to="$(arg image_info_topic)" />
    <remap from="/device_0/sensor_0/Depth_0/image/data" to="$(arg depth_topic)" />
    <remap from="/device_0/sensor_0/Depth_0/info/camera_info" to="$(arg depth_info_topic)" /> -->
    <!-- <remap from="/device_0/sensor_0/Depth_0/depth/data" to="/$(arg used_camera)/depth/image_raw" /> -->
    <!-- <remap from="/device_0/sensor_0/Depth_0/info/camera_info" to="/$(arg used_camera)/depth/camera_info" /> -->

    <!-- /clockがないにもかかわらずuse_sim_timeがtrueだとrospy.sleepでブロックがかかる ref: https://github.com/ros/ros_comm/issues/2111 -->
    <!-- <param name="use_sim_time" value="true" /> -->
    <param name="use_sim_time" value="false" />

    <param name="fps" value="$(arg fps)" />
    <param name="device" value="$(arg device)" />

    <param name="finger_num" value="$(arg finger_num)" />
    <param name="unit_angle" value="$(arg unit_angle)" />
    <param name="hand_radius_mm" value="$(arg hand_radius_mm)" />
    <param name="finger_radius_mm" value="$(arg finger_radius_mm)" />
    <param name="hand_mount_rotation" value="$(arg hand_mount_rotation)" />
    
    <param name="robot_name" value="$(arg robot_name)" />
    <param name="debug" value="$(arg debug)" />
    <param name="visualize_only_best_cnd" value="$(arg visualize_only_best_cnd)" />
    
    <!-- <param name="image_topic" value="$(arg image_topic)" />
    <param name="depth_topic" value="$(arg depth_topic)" /> -->
    <param name="image_info_topic" value="$(arg image_info_topic)" />
    <!-- <param name="depth_info_topic" value="$(arg depth_info_topic)" />
    <param name="points_topic" value="$(arg points_topic)" /> -->
    <param name="seg_topic" value="$(arg seg_topic)" />
    <!-- <param name="candidates_img_topic" value="$(arg candidates_img_topic)" /> -->

    <!-- <node name="raw_view" pkg="rqt_image_view" type="rqt_image_view" args="/body_camera/color/image_raw" /> -->
    <group if="$(arg show_view)">
        <!-- <node name="depth_view" pkg="rqt_image_view" type="rqt_image_view" args="$(arg depth_topic)" /> -->
        <node name="seg_view" pkg="rqt_image_view" type="rqt_image_view" args="$(arg seg_topic)" />
        <node name="grasp_view" pkg="rqt_image_view" type="rqt_image_view" args="/grasp_detection_server_result" />
    </group>

    <group if="$(arg use_bag)">
        <node name="rosbag_play" pkg="rosbag" type="play" args="$(arg bag) -r $(arg fps) -l --clock" />
        <node name="grasp_detection_test_client" pkg="detect" type="test_grasp_detection.py" output="screen" />
    </group>

    <!-- action server -->
    <node name="tf_transform_server" pkg="detect" type="tf_transform_server.py" />
    <node name="instance_segmentation_server" pkg="detect" type="instance_segmentation_server.py" />
    <node name="grasp_detection_server" pkg="detect" type="grasp_detection_server.py" output="screen" />
    <node name="container_position_server" pkg="detect" type="container_position_server.py" />
    <node name="exclusion_list_server" pkg="detect" type="exclusion_list_server.py" />
</launch>